src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,752 @@
+    }
+
+    // Case 1B: There is a conflict with a group from a different nontrivial
+    // equivalnece class. Because we canonicalize the shapes whenever we first
+    // create a nontrivial equivalence class, this is usually not possible
+    // because two equivalence classes with isomorphic types should actually be
+    // the same equivalence class. The exception is when the base shapes of two
+    // different equivalence classes are the same, but there is still a conflict
+    // due to brands in each class matching the base type of the other class. In
+    // this case we don't actually want to join the classes; we just want to
+    // advance to the next configuration to resolve the conflict.
+    if (groups[groupRep].classInfo && groups[otherRep].classInfo) {
+      auto& classInfo = *groups[groupRep].classInfo;
+      classInfo.advance();
+      classInfo.permute(groupInfo);
+      shapesToUpdate.push_back(group);
+      return;
+    }
+
+    // Case 1C: We have permuted ourselves into equivalence with some other
+    // group unaffiliated with a nontrivial equivalence class. Bring that other
+    // group into our equivalence class and try the next permutation instead.
+    if (groups[groupRep].classInfo) {
+      auto& classInfo = *groups[groupRep].classInfo;
+
+      [[maybe_unused]] size_t unionRep =
+        equivalenceClasses.getUnion(groupRep, otherRep);
+      assert(groupRep == unionRep);
+
+      // `other` must have the same permutation as `group` because they have the
+      // same shape. Advance `group` to the next permutation.
+      otherInfo.classInfo = std::nullopt;
+      otherInfo.permutation = groupInfo.permutation;
+      classInfo.advance();
+      classInfo.permute(groupInfo);
+
+      shapesToUpdate.push_back(group);
+      return;
+    }
+
+    // Case 2A: This shape is already found in an existing nontrivial
+    // equivalence class. Join the class and try the next permutation.
+    if (groups[otherRep].classInfo) {
+      auto& classInfo = *groups[otherRep].classInfo;
+
+      [[maybe_unused]] size_t unionRep =
+        equivalenceClasses.getUnion(otherRep, groupRep);
+      assert(otherRep == unionRep);
+
+      // Since we matched shapes with `other`, unapplying its permutation gets
+      // us back to the canonical shape, from which we can apply a fresh
+      // permutation.
+      groupInfo.classInfo = std::nullopt;
+      groupInfo.permutation = otherInfo.permutation;
+      classInfo.advance();
+      classInfo.permute(groupInfo);
+
+      shapesToUpdate.push_back(group);
+      return;
+    }
+
+    // Case 2B: We need to join two matching unaffiliated groups into a new
+    // nontrivial equivalence class.
+    assert(!groups[groupRep].classInfo && !groups[otherRep].classInfo);
+    assert(group == groupRep && other == otherRep);
+
+    // We are canonicalizing and reinserting the shape for other, so remove it
+    // from the map under its current shape.
+    groupShapeIndices.erase(it);
```

Comment:
`it` is for `group`, not `other`, so I am confused how to map the comment to the code here?

Reply:
No, `it` is for `other` because it is the iterator to the existing element we found when we tried to insert `group`.

Reply:
Ah, right, thanks!